src/Uno.Toolkit.UI/Controls/Shadows/ShadowContainer.Paint.cs

```diff
@@ -0,0 +1,382 @@
+﻿using System;
+using System.Linq;
+
+using SkiaSharp;
+
+using Windows.UI;
+
+using Microsoft.Extensions.Logging;
+using Uno.Extensions;
+using Uno.Logging;
+
+#if IS_WINUI
+using Microsoft.UI.Xaml.Media;
+using Microsoft.UI.Xaml.Controls;
+using SkiaSharp.Views.Windows;
+#else
+using Windows.UI.Xaml.Media;
+using Windows.UI.Xaml.Controls;
+using SkiaSharp.Views.UWP;
+#endif
+
+namespace Uno.Toolkit.UI;
+
+public partial class ShadowContainer
+{
+	private static readonly ILogger _logger = typeof(InputExtensions).Log();
+
+	private record ShadowInfos(double Width, double Height, bool IsInner, double BlurRadius, double Spread, double X, double Y, Color color)
+	{
+		public static readonly ShadowInfos Empty = new ShadowInfos(0, 0, false, 0, 0, 0, 0, new Color());
+
+		public static ShadowInfos From(Shadow shadow, double width, double height)
+		{
+			return new ShadowInfos(
+				width,
+				height,
+				shadow.IsInner,
+				shadow.BlurRadius,
+				shadow.Spread,
+				shadow.OffsetX,
+				shadow.OffsetY,
+				Color.FromArgb((byte)(shadow.Color.A * shadow.Opacity), shadow.Color.R, shadow.Color.G, shadow.Color.B));
+		}
+	}
+
+	private record struct SKShadow(
+		bool IsInner,
+		float OffsetX,
+		float OffsetY,
+		float BlurSigma,
+		SKColor Color,
+		float Spread,
+		float ContentWidth,
+		float ContentHeight,
+		float CornerRadius)
+	{
+		public static SKShadow From(Shadow shadow, float width, float height, float cornerRadius, float pixelRatio)
+		{
+			float blurRadius = (float)shadow.BlurRadius * pixelRatio;
+			// Blur sigma conversion taken from flutter source code
+			float blurSigma = blurRadius > 0 ? blurRadius * 0.57735f + 0.5f : 0f;
+
+			// Can't use ToSKColor() or we end up with a weird compilation error asking us to reference System.Drawing
+			Color windowsUiColor = shadow.Color;
+			var color = ToSkiaColor(windowsUiColor);
+			color = color.WithAlpha((byte)(color.Alpha * shadow.Opacity));
+
+			return new SKShadow(
+				shadow.IsInner,
+				(float)shadow.OffsetX * pixelRatio,
+				(float)shadow.OffsetY * pixelRatio,
+				blurSigma,
+				color,
+				(float)shadow.Spread * pixelRatio,
+				width,
+				height,
+				cornerRadius);
+		}
+	}
+
+	private ShadowInfos[] _shadowInfoArray = new[] { ShadowInfos.Empty };
+	private float _currentPixelRatio;
+	private Color? _currentContentBackgroundColor;
+
+	private bool NeedsPaint(double width, double height, float pixelRatio, out bool pixelRatioChanged)
+	{
+		var shadows = Shadows ?? new ShadowCollection();
+		var newShadowInfos = shadows.Select(s => ShadowInfos.From(s, width, height)).ToArray();
+
+		pixelRatioChanged = false;
+
+		bool needsPaint = !newShadowInfos.SequenceEqual(_shadowInfoArray);
+		_shadowInfoArray = newShadowInfos;
+
+		if (pixelRatio != _currentPixelRatio)
+		{
+			_currentPixelRatio = pixelRatio;
+			pixelRatioChanged = needsPaint = true;
+		}
+
+		return needsPaint;
+	}
+
+#if false // ANDROID  (see comment in ShadowContainer.cs)
```

Comment:
Do we need to keep this?

Reply:
I created this issue about that: https://github.com/unoplatform/uno.toolkit.ui/issues/694